DOC Notebook style and enhanced descriptions in SVM kernels example

StefanieSenger opened the pull request:
#### What does this implement/fix? Explain your changes.
This PR suggests to introduce notebook style formatting and enhanced descriptions in the [SVM kernels example](https://scikit-learn.org/stable/auto_examples/svm/plot_svm_kernels.html#sphx-glr-auto-examples-svm-plot-svm-kernels-py), 

#### Any other comments?
A few issues:

1. The math formulas break when also applying a line break. I couldn't find how to fix that.
2.  I'd like to include the sigmoid kernel later, but need to do some reading on it first.

Looking forward to your comments. :)

StefanieSenger committed:
DOC Notebook style and enhanced descriptions in SVM kernels example

adrinjalali commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,285 @@
+#
+# RBF kernel
+# **********
+# The radial basis function (RBF) kernel, also known as the Gaussian kernel, is
+# the default kernel for Support Vector Machines in scikit-learn. It measures
+# similarity between two data points in infinite dimensions and then approaches
+# classification by majority vote. The kernel function is defined as:
```
```suggestion
# The radial basis function (RBF) kernel, also known as the Gaussian kernel, is
# the default kernel for Support Vector Machines in scikit-learn. It measures
# similarity between two data points in infinite dimensions and then approaches
# classification by majority vote. The kernel function is defined as:
```

I don't think we can give an intuition really here.

StefanieSenger committed:
added description for sigmoid kernel

StefanieSenger committed:
Merge remote-tracking branch 'upstream/main' into examples_svm_kernels

adrinjalali reviewed (commented):
I think you can fix the math with this patch:


```patch
diff --git a/examples/svm/plot_svm_kernels.py b/examples/svm/plot_svm_kernels.py
index 3c4772fa3..e475d6936 100644
--- a/examples/svm/plot_svm_kernels.py
+++ b/examples/svm/plot_svm_kernels.py
@@ -210,8 +210,8 @@ plot_training_data_with_decision_boundary("linear")
 # The polynomial kernel changes the notion of similarity. The kernel function
 # is defined as:
 #
-# .. math:: K(\mathbf{x}_1, \mathbf{x}_2) = (\gamma \cdot \
-# .. math:: \mathbf{x}_1^\top\mathbf{x}_2 + r)^d
+# .. math::
+#   K(\mathbf{x}_1, \mathbf{x}_2) = (\gamma \cdot \mathbf{x}_1^\top\mathbf{x}_2 + r)^d
 #
 # where :math:`{d}` is the degree (`degree`) of the polynomial, :math:`{\gamma}`
 # (`gamma`) controls the influence of each individual training sample on the
@@ -238,8 +238,9 @@ plot_training_data_with_decision_boundary("poly")
 # similarity between two data points in infinite dimensions and then approaches
 # classification by majority vote. The kernel function is defined as:
 #
-# .. math:: K(\mathbf{x}_1, \mathbf{x}_2) = \exp\left(-\gamma \cdot \
-# .. math:: {\|\mathbf{x}_1 - \mathbf{x}_2\|^2}\right)
+# .. math::
+#   K(\mathbf{x}_1, \mathbf{x}_2) = \exp\left(-\gamma \cdot \
+#       {\|\mathbf{x}_1 - \mathbf{x}_2\|^2}\right)
 #
 # where :math:`{\gamma}` (`gamma`) controls the influence of each individual
 # training sample on the decision boundary.
@@ -258,8 +259,9 @@ plot_training_data_with_decision_boundary("rbf")
 # **************
 # The sigmoid kernel function is defined as:
 #
-# .. math:: K(\mathbf{x}_1, \mathbf{x}_2) = \tanh(\gamma \cdot \\
-# .. math:: \mathbf{x}_1^\top\mathbf{x}_2 + r)
+# .. math::
+#   K(\mathbf{x}_1, \mathbf{x}_2) = \tanh(\gamma \cdot \
+#       \mathbf{x}_1^\top\mathbf{x}_2 + r)
 #
 plot_training_data_with_decision_boundary("sigmoid")
```

StefanieSenger committed:
changes after review

adrinjalali reviewed (approved):
@glemaitre do you wanna have a look?

adrinjalali commented:
@StefanieSenger could you please also add a link to this example in the SVM api docs?

adrinjalali commented:
The issue with the ones automatically shown on the bottom of API docs is that it's really not clear what example is relevant for what. Adding the link explicitly in the API doc would make these much more discoverable. I'd add it to the docstring of `kernel` parameter, or under the `Examples` section.

glemaitre commented:
> @glemaitre do you wanna have a look?

Yep I will have a look now.

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,306 @@
+# Plotting settings
+fig, ax = plt.subplots(figsize=(4, 3))
+ax.grid(True)
+x_min = -3
+x_max = 3
+y_min = -3
+y_max = 3
```
```suggestion
x_min, xmax, y_min, y_max = -3, 3, -3, 3
```

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,306 @@
+y_max = 3
+plt.xlim(x_min, x_max)
+plt.ylim(y_min, y_max)
+samples_colormap = np.array(["b", "r"])
```
@adrinjalali does not seem to be the best choice for color blindness, ins't it?
The original violet/yellow should bring more constrast here (it means only passing `c=y`)

adrinjalali replied:
blue and red are actually quite okay, it's red and green which is the issue. I don't mind the colors as they are, but also like the colors as they were.

StefanieSenger replied:
Okay, let's keep the new colors then. :)

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,306 @@
+x_max = 3
+y_min = -3
+y_max = 3
+plt.xlim(x_min, x_max)
+plt.ylim(y_min, y_max)
```
```suggestion
ax.set_xlim(x_min, x_max)
ax.set_ylim(y_min, y_max)
```

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,306 @@
-    )
-    plt.scatter(X[:, 0], X[:, 1], c=Y, zorder=10, cmap=plt.cm.Paired, edgecolors="k")
+# Plot samples by color and add legend
+ax.scatter(X[:, 0], X[:, 1], s=150, c=samples_colormap[y], label=y, edgecolors="k")
+legend_labels = ["class 0", "class 1"]
+legend_handles = [
+    plt.Line2D([0], [0], marker="o", color="w", markerfacecolor=color, markersize=10)
+    for color in samples_colormap
+]
+ax.legend(legend_handles, legend_labels)
```
I think that we can reduce the boilerplate and have a fine results.
```suggestion
scatter = ax.scatter(X[:, 0], X[:, 1], s=150, c=samples_colormap[y], label=y, edgecolors="k")
ax.legend(*scatter.legend_elements(), loc="lower right", title="Classes")
```

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,306 @@
+
+def plot_training_data_with_decision_boundary(kernel):
+    # Train the SVC
+    clf = svm.SVC(kernel=kernel, gamma=2)
+    clf.fit(X, y)
```
```suggestion
    clf = svm.SVC(kernel=kernel, gamma=2).fit(X, y)
```

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,306 @@
+    # Plotting settings
+    fig, ax = plt.subplots(figsize=(4, 3))
+    ax.grid(True)
     x_min = -3
     x_max = 3
     y_min = -3
     y_max = 3
```
```suggestion
    x_min, x_max, y_min, y_max = -3, 3, -3, 3
```

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,306 @@
     x_max = 3
     y_min = -3
     y_max = 3
+    plt.xlim(x_min, x_max)
+    plt.ylim(y_min, y_max)
```
```suggestion
    ax.set_xlim(x_min, x_max)
    ax.set_ylim(y_min, y_max)
```

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,306 @@
+    )
+    # Plot samples by color and add legend
+    ax.scatter(X[:, 0], X[:, 1], c=samples_colormap[y], s=150, edgecolors="k")
+    legend_labels = ["class 0", "class 1"]
+    legend_handles = [
+        plt.Line2D(
+            [0], [0], marker="o", color="w", markerfacecolor=color, markersize=10
+        )
+        for color in samples_colormap
+    ]
+    ax.legend(legend_handles, legend_labels)
```
We can simplify the legend

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,306 @@
+    # Extract the decision boundary
     XX, YY = np.mgrid[x_min:x_max:200j, y_min:y_max:200j]
     Z = clf.decision_function(np.c_[XX.ravel(), YY.ravel()])
 
     # Put the result into a color plot
     Z = Z.reshape(XX.shape)
-    plt.figure(fignum, figsize=(4, 3))
-    plt.pcolormesh(XX, YY, Z > 0, cmap=plt.cm.Paired)
+    plt.pcolormesh(XX, YY, Z > 0, cmap=ListedColormap(["lightblue", "lightpink"]))
     plt.contour(
         XX,
         YY,
         Z,
         colors=["k", "k", "k"],
         linestyles=["--", "-", "--"],
-        levels=[-0.5, 0, 0.5],
+        levels=[-1, 0, 1],
     )
```
We can use our own plotting tools here. At the end it will not be mush different but it hide the `meshgrid`:

```python
       from sklearn.inspection import DecisionBoundaryDisplay

    common_params = {
        "estimator": clf, "X": X, "ax": ax, "response_method": "decision_function"
    }
    DecisionBoundaryDisplay.from_estimator(**common_params, plot_method="pcolormesh")
    DecisionBoundaryDisplay.from_estimator(
        **common_params,
        plot_method="contour",
        levels=[-1, 0, 1],
        colors=["k", "k", "k"],
        linestyles=["--", "-", "--"],
    )
```

glemaitre reviewed (commented):
I propose some simplification regarding the plotting part.
I will check the narrative in early afternoon.

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,306 @@
-    plt.clf()
+# Plotting settings
+fig, ax = plt.subplots(figsize=(4, 3))
+ax.grid(True)
```
Let's remove the grid. It will be useless in the second part when we have the decision function

glemaitre reviewed (commented):
I am fine with the rest of the example.

StefanieSenger committed:
changes after review

StefanieSenger commented:
Thanks for your review @glemaitre. I have substituted the old matplotlib plotting by our `DecisionBoundaryDisplay` and made some other small adjustments that you suggested. :)

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,301 @@
+cl1_label = plt.Line2D(
+    [], [], marker="o", color="black", markerfacecolor="r", markersize=10
+)
+ax.legend([cl0_label, cl1_label], ["Class 0", "Class 1"], loc="lower right")
```
```suggestion
ax.legend(*scatter.legend_elements(), loc="upper right", title="Classes")
```

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,301 @@
+# In a real classification task, where performance matters, parameter tuning
+# with techniques like :class:`~sklearn.model_selection.GridSearchCV` is
+# highly recommended to capture different structures within the data.
+from matplotlib.colors import ListedColormap
+
```
```suggestion
```

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,301 @@
+    clf = svm.SVC(kernel=kernel, gamma=2).fit(X, y)
+
+    # Settings for plotting
+    fig, ax = plt.subplots(figsize=(4, 3))
```
```suggestion
    _, ax = plt.subplots(figsize=(4, 3))
```

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,301 @@
+    # Settings for plotting
+    fig, ax = plt.subplots(figsize=(4, 3))
+    x_min, x_max, y_min, y_max = -3, 3, -3, 3
+    ax.set_xlim(x_min, x_max)
+    ax.set_ylim(y_min, y_max)
+    samples_colormap = np.array(["b", "r"])
```
```suggestion
    ax.set(xlim=(x_min, x_max), ylim=(y_min, y_max))
```

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,301 @@
+        **common_params,
+        response_method="predict",
+        plot_method="pcolormesh",
+        cmap=ListedColormap(["lightblue", "lightpink"]),
```
```suggestion
        alpha=0.3,
```

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,301 @@
-        levels=[-0.5, 0, 0.5],
+    # Plot samples by color and add legend
+    ax.scatter(X[:, 0], X[:, 1], c=samples_colormap[y], s=150, edgecolors="k")
+    class0_label = plt.Line2D(
+        [], [], marker="o", color="black", markerfacecolor="b", markersize=10
+    )
+    class1_label = plt.Line2D(
+        [], [], marker="o", color="black", markerfacecolor="r", markersize=10
     )
```
You can remove those lines

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,301 @@
-        linestyles=["--", "-", "--"],
-        levels=[-0.5, 0, 0.5],
+    # Plot samples by color and add legend
+    ax.scatter(X[:, 0], X[:, 1], c=samples_colormap[y], s=150, edgecolors="k")
```
```suggestion
    ax.scatter(X[:, 0], X[:, 1], c=y, s=150, edgecolors="k")
```

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,301 @@
+    class1_label = plt.Line2D(
+        [], [], marker="o", color="black", markerfacecolor="r", markersize=10
     )
+    ax.legend([class0_label, class1_label], ["Class 0", "Class 1"], loc="lower right")
```
```suggestion
    ax.legend(*scatter.legend_elements(), loc="upper right", title="Classes")
```

glemaitre commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,301 @@
-    plt.yticks(())
-    fignum = fignum + 1
-plt.show()
+# %%
```
```suggestion
```

glemaitre reviewed (approved):
Just a couple of matplotlib simplifications.

Otherwise LGTM.

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

StefanieSenger committed:
changes after review

StefanieSenger commented:
Thank you @glemaitre for editing the plotting and the suggestions. :)

ArturoAmorQ commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,279 @@
+plot_training_data_with_decision_boundary("linear")
+
+# %%
+# Training a :class:`~sklearn.svm.SVC` on a linear kernel results in an
+# untransformed feature space and we can see the hyperplane and the margins as
+# straight lines. The support vectors are on the margins. The trained classes do
+# not perfectly capture the data, but thanks to the model's soft margins, we can
+# expect that it has converged, and the position of the classification
+# boundaries has stabilized.
```
I moved here the line concerning the choice `C=1` resulting in margins at 1 unit distance, as this seems to be true only for the linear kernel, but it would be nice if you could add a reference or the reason why that is the case.
```suggestion
# Training a :class:`~sklearn.svm.SVC` on a linear kernel results in an
# untransformed feature space where the hyperplane and the margins are straight
# lines. In this case, the choice `C=1` results in margins at 1 unit distance
# from both sides of the hyperplane. Due to the lack of expressivity of the
# linear kernel, the trained classes do not perfectly predict the data.
```

ArturoAmorQ reviewed (commented):
Thanks for the PR @StefanieSenger! I suggest some wording.

May you also please modify the `supervised_learning.rst` file to keep the intended order of the images? And as mentioned in RL, you can also add a small section on such file referring the new sigmoid kernel image.

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

StefanieSenger commented:
Thank you, @ArturoAmorQ for your comments. You are right with your procaution regarding the margins being at a distance from the boundary equal to the C param on the linear kernel. I was assuming something empirically, that might as well have been coincidence. I have deleted that information to be sure.

I have modified the `supervised_learning.rst ` file, that refers to the plots, and added the sigmoid plot there, too.

Unfortunately I have pushed some wrong files along with the changes (was late yesterday). I will take care of this, but I am not sure if I have time for this before Wednesday. I have turned this PR into a draft in the meantime.

StefanieSenger committed:
changes after review

ArturoAmorQ committed:
Merge branch 'main' into examples_svm_kernels

ArturoAmorQ commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,272 @@
+#   K(\mathbf{x}_1, \mathbf{x}_2) = \tanh(\gamma \cdot
+#       \mathbf{x}_1^\top\mathbf{x}_2 + r)
+#
+# where :math:`{\gamma}` (`gamma`) controls the influence of each individual
+# training sample on the decision boundary and :math:`{r}` is the bias term
+# (`coef0`) that shifts the data up or down.
+#
+# In the sigmoid kernel, the similarity between two data points is computed
+# using the hyperbolic tangent function (tanh). The kernel function takes into
+# account the dot product of the two points (x1 and x2), scaled by the kernel
+# coefficient γ, and shifted by a bias term r.
```
I suggest merging these two paragraphs into a single one to avoid redundancies:

```suggestion
# where the similarity between two data points is computed using the hyperbolic
# tangent function (:math:`tanh`). The kernel function takes into account the
# dot product of the two points (:math:`\mathbf{x}_1` and :math:`\mathbf{x}_2`),
# scaled by the kernel coefficient :math:`{\gamma}`, and shifted by a bias term
# :math:`{r}` (`coef0`).
```

StefanieSenger replied:
I'd rather leave the paragraph explaining the components of the mathematical formula separate, directly following the formula like in the other paragraphs, if this is alright. But I have improved the paragraph with :math: notation. :)

ArturoAmorQ commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,272 @@
+# quality. They are intended to provide a visual understanding of how the
+# different kernels use the training data.
+#
+# For a comprehensive evaluation, fine-tuning of SVC parameters using techniques
+# such as :class:`~sklearn.model_selection.GridSearchCV` is recommended to
+# capture the underlying structures within the data.
```
```suggestion
# For a comprehensive evaluation, fine-tuning of :class:`~sklearn.svm.SVC`
# parameters using techniques such as
# :class:`~sklearn.model_selection.GridSearchCV` is recommended to capture the
# underlying structures within the data.
```

ArturoAmorQ reviewed (commented):
Just some final nitpicks :)

ArturoAmorQ commented on sklearn/svm/_classes.py:
```diff
@@ -918,7 +920,9 @@ class NuSVC(BaseSVC):
         default='rbf'
          Specifies the kernel type to be used in the algorithm.
          If none is given, 'rbf' will be used. If a callable is given it is
-         used to precompute the kernel matrix.
+         used to precompute the kernel matrix. For an intuitive
+        visualization of different kernel types see
```
The CI is failing because of the indentation here, which was not your fault. Can you please remove one blank space from current lines 921 to 923?

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

StefanieSenger committed:
changes after review

StefanieSenger committed:
Merge branch 'examples_svm_kernels' of github.com:StefanieSenger/scikit-learn into examples_svm_kernels

ArturoAmorQ commented on examples/svm/plot_svm_kernels.py:
```diff
@@ -1,93 +1,274 @@
+#   K(\mathbf{x}_1, \mathbf{x}_2) = \tanh(\gamma \cdot
+#       \mathbf{x}_1^\top\mathbf{x}_2 + r)
+#
+# where :math:`{\gamma}` (`gamma`) controls the influence of each individual
+# training sample on the decision boundary and :math:`{r}` is the bias term
+# (`coef0`) that shifts the data up or down.
+#
+# In the sigmoid kernel, the similarity between two data points is computed
+# using the hyperbolic tangent function (:math:`tanh`). The kernel function
+# takes into account the dot product of the two points (:math:`\mathbf{x}_1`
+# and :math:`\mathbf{x}_2`), scaled by the kernel coefficient :math:`{\gamma}`
+# (`gamma`), and shifted by a bias term :math:`{r}` (`coef0`).
```
I still think that the kernel coefficient and the bias are defined two times. WDYT about the following suggestion?

```suggestion
# where the kernel coefficient :math:`{\gamma}` (`gamma`) controls the influence
# of each individual training sample on the decision boundary and :math:`{r}` is
# the bias term (`coef0`) that shifts the data up or down.
#
# In the sigmoid kernel, the similarity between two data points is computed
# using the hyperbolic tangent function (:math:`\tanh`). The kernel function
# scales and possibly shifts the dot product of the two points
# (:math:`\mathbf{x}_1` and :math:`\mathbf{x}_2`).
```

StefanieSenger replied:
I'm fine with that. Have committed your suggestion.

StefanieSenger committed:
Update examples/svm/plot_svm_kernels.py

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

ArturoAmorQ reviewed (approved):
LGTM, thanks for the PR @StefanieSenger! :tada:

On a side note, this is going to be my very first merge on scikit-learn :)

StefanieSenger commented:
> On a side note, this is going to be my very first merge on scikit-learn :)

Yeahhhh, let's celebrate, @ArturoAmorQ  🎉🎉🎉
